### PR TITLE
small improvements for parachains consensus

### PR DESCRIPTION
**Pull request by rphmeier**
* add a delay before proposing so we wait for candidates & bitfields
* add logging of reputation changes under `network_bridge=debug`
* provisioner: waiting for timeout doesn't block message processing

### CODE REVIEW

File: node/core/provisioner/src/lib.rs [hunk 1]
```diff
@@ -126,70 +168,90 @@ impl ProvisioningJob {
...
+				msg = self.receiver.next().fuse() => match msg {
+					Some(RequestInherentData(_, return_sender)) => {
+						let _timer = self.metrics.time_request_inherent_data();
+
+						if self.inherent_after.is_ready() {
+							self.send_inherent_data(vec![return_sender]).await;
+						} else {
+							self.awaiting_inherent.push(return_sender);
 						}
+
```
Comment (bkchr) [inline node/core/provisioner/src/lib.rs:181]:
```suggestion
```

File: node/core/provisioner/src/lib.rs [hunk 2]
```diff
@@ -126,70 +168,90 @@ impl ProvisioningJob {
...
+					}
+					Some(ProvisionableData(_, data)) => {
+						let _timer = self.metrics.time_provisionable_data();
+
+						let mut bad_indices = Vec::new();
+						for (idx, channel) in self.provisionable_data_channels.iter_mut().enumerate() {
+							match channel.send(data.clone()).await {
+								Ok(_) => {}
+								Err(_) => bad_indices.push(idx),
 							}
```
Comment (bkchr) [inline node/core/provisioner/src/lib.rs:194]:
```suggestion
							if channel.send(data.clone()).await.is_err() {
								bad_indices.push(idx),
							}
```

File: node/core/provisioner/src/lib.rs [hunk 3]
```diff
@@ -126,70 +168,90 @@ impl ProvisioningJob {
...
+								}
+								retain
+							})
+							.map(|(_, item)| item)
+							.collect();
+					}
+					None => break,
+				},
+				_ = self.inherent_after.ready().fuse() => {
+					let return_senders = std::mem::replace(&mut self.awaiting_inherent, Vec::new());
```
Comment (bkchr) [inline node/core/provisioner/src/lib.rs:224]:
```suggestion
					let return_senders = std::mem::take(&mut self.awaiting_inherent);
```

File: node/core/provisioner/src/lib.rs [hunk 4]
```diff
@@ -126,70 +168,90 @@ impl ProvisioningJob {
...
+	async fn send_inherent_data(
+		&mut self,
+		return_senders: Vec<oneshot::Sender<ProvisionerInherentData>>,
+	) {
+		if let Err(err) = send_inherent_data(
+			self.relay_parent,
+			&self.signed_bitfields,
+			&self.backed_candidates,
+			return_senders,
+			self.sender.clone(),
```
Comment (bkchr) [inline node/core/provisioner/src/lib.rs:244]:
```suggestion
			&mut self.sender,
```

File: node/core/provisioner/src/lib.rs [hunk 5]
```diff
@@ -223,12 +285,12 @@ type CoreAvailability = BitVec<bitvec::order::Lsb0, u8>;
...
 /// choose a coherent set of candidates along with that.
-#[tracing::instrument(level = "trace", skip(return_sender, from_job), fields(subsystem = LOG_TARGET))]
+#[tracing::instrument(level = "trace", skip(return_senders, from_job), fields(subsystem = LOG_TARGET))]
 async fn send_inherent_data(
 	relay_parent: Hash,
 	bitfields: &[SignedAvailabilityBitfield],
 	candidates: &[BackedCandidate],
-	return_sender: oneshot::Sender<ProvisionerInherentData>,
+	return_senders: Vec<oneshot::Sender<ProvisionerInherentData>>,
 	mut from_job: mpsc::Sender<FromJobCommand>,
```
Comment (bkchr) [inline node/core/provisioner/src/lib.rs:294]:
```suggestion
	from_job: &mut mpsc::Sender<FromJobCommand>,
```

File: node/core/provisioner/src/lib.rs [hunk 6]
```diff
@@ -223,12 +285,12 @@ type CoreAvailability = BitVec<bitvec::order::Lsb0, u8>;
...
 ) -> Result<(), Error> {
 	let availability_cores = request_availability_cores(relay_parent, &mut from_job)
```
Comment (bkchr) [inline node/core/provisioner/src/lib.rs:296]:
```suggestion
	let availability_cores = request_availability_cores(relay_parent, from_job)
```

Comment (bkchr) [reply node/core/provisioner/src/lib.rs:296]: Line 306 also needs to be changed.

File: node/core/provisioner/src/lib.rs [hunk 7]
```diff
@@ -38,9 +38,42 @@ use polkadot_primitives::v1::{
 };
 use std::{pin::Pin, collections::BTreeMap};
 use thiserror::Error;
+use futures_timer::Delay;
+
+/// How long to wait before proposing,.
```
Comment (bkchr) [inline node/core/provisioner/src/lib.rs:43]:
```suggestion
/// How long to wait before proposing.
```

**Review by bkchr — APPROVED** [review]

File: node/core/provisioner/src/lib.rs (hunk 4 above)
Comment (rphmeier) [reply node/core/provisioner/src/lib.rs:244]: This interferes with the `Send` implementation - i tried this locally and got compilation errors

**Comment by rphmeier** [comment]
bot merge

**Comment by bkchr** [comment]
bot merge

**Comment by bkchr** [comment]
bot merge

**Pull request merged**